security/advancedtls/pemfile_provider.go

```diff
@@ -0,0 +1,179 @@
+type PEMFileProvider struct {
+	identityDistributor *certprovider.Distributor
+	rootDistributor     *certprovider.Distributor
+	cancel              context.CancelFunc
+}
+
+// NewPEMFileProvider uses PEMFileProviderOptions to construct a PEMFileProvider.
+func NewPEMFileProvider(o *PEMFileProviderOptions) (*PEMFileProvider, error) {
+	if o.CertFile == "" && o.KeyFile == "" && o.TrustFile == "" {
+		return nil, fmt.Errorf("at least one credential file needs to be specified")
+	}
+	if keySpecified, certSpecified := o.KeyFile != "", o.CertFile != ""; keySpecified != certSpecified {
+		return nil, fmt.Errorf("private key file and identity cert file should be both specified or not specified")
+	}
+	if o.IdentityInterval == nil {
+		o.IdentityInterval = &defaultIdentityInterval
+	}
+	if o.RootInterval == nil {
+		o.RootInterval = &defaultRootInterval
+	}
+	provider := &PEMFileProvider{}
+	if o.CertFile != "" && o.KeyFile != "" {
+		provider.identityDistributor = certprovider.NewDistributor()
+	}
+	if o.TrustFile != "" {
+		provider.rootDistributor = certprovider.NewDistributor()
+	}
+	// A goroutine to pull file changes.
+	identityTicker := time.NewTicker(*o.IdentityInterval)
+	rootTicker := time.NewTicker(*o.RootInterval)
+	ctx, cancel := context.WithCancel(context.Background())
+	// We pass a copy of PEMFileProviderOptions to the goroutine in case users change it after we start reloading.
+	go func(ctx context.Context, o PEMFileProviderOptions) {
+		for {
+			select {
+			case <-ctx.Done():
+				identityTicker.Stop()
+				rootTicker.Stop()
+				return
+			case <-identityTicker.C:
+				if provider.identityDistributor == nil {
+					continue
+				}
+				// Read identity certs from PEM files.
+				identityCert, err := readKeyCertPairFunc(o.CertFile, o.KeyFile)
+				if err != nil {
+					// If the reading produces an error, we will skip the update for this round and log the error.
+					logger.Warningf("tls.LoadX509KeyPair reads %s and %s failed: %v", o.CertFile, o.KeyFile, err)
+					continue
+				}
+				provider.identityDistributor.Set(&certprovider.KeyMaterial{Certs: []tls.Certificate{identityCert}}, nil)
+			case <-rootTicker.C:
+				if provider.rootDistributor == nil {
+					continue
+				}
+				// Read root certs from PEM files.
+				trustPool, err := readTrustCertFunc(o.TrustFile)
+				if err != nil {
+					// If the reading produces an error, we will skip the update for this round and log the error.
+					logger.Warningf("readTrustCertFunc reads %v failed: %v", o.TrustFile, err)
+					continue
+				}
+				provider.rootDistributor.Set(&certprovider.KeyMaterial{Roots: trustPool}, nil)
+			default:
+			}
+		}
+	}(ctx, *o)
+	provider.cancel = cancel
+	return provider, nil
+}
+
+// KeyMaterial returns the key material sourced by the PEMFileProvider.
+// Callers are expected to use the returned value as read-only.
+func (p *PEMFileProvider) KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error) {
+	km := certprovider.KeyMaterial{}
```

Comment:
Defining this as a pointer here will make the return statement look cleaner.
Here:
`km := &certprovider.KeyMaterial{}`
Return:
`return km, nil`